errors/error_test.go

```diff
@@ -19,6 +20,35 @@ const (
 
 type testContextKey string
 
+func TestNew(t *testing.T) {
+	childErr := New(Op(childOp), BadRequest, childErrMsg)
+
+	// Test for data race condition when trying to `e.propagateContexts()`.
+	var wg sync.WaitGroup
+	for i := 0; i < 2; i++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			err := New(Op(parentOp), parentErrMsg, childErr)
+			wantErr := &Error{
+				Op:  parentOp,
+				msg: parentErrMsg,
+				Err: &Error{
+					Op:   childOp,
+					Kind: BadRequest,
+					msg:  childErrMsg,
+				},
+			}
+
+			assert.Equal(t, wantErr, err)
+		}()
+	}
+
+	wg.Wait()
+}
```

Comment:
This test allows us to catch race condition. Done in the Test Driven Development manner, the test was failing:

```sh
$ make test
go test ./... -race -cover -p 5
==================
WARNING: DATA RACE
Read at 0x00c0000bcff8 by goroutine 19:
  github.com/wego/pkg/errors.(*Error).basics()
      /go/src/github.com/wego/pkg/errors/error.go:206 +0x118
  github.com/wego/pkg/errors.(*Error).propagateContexts()
      /go/src/github.com/wego/pkg/errors/error.go:175 +0xac
  github.com/wego/pkg/errors.New()
      /go/src/github.com/wego/pkg/errors/error.go:77 +0x478
  github.com/wego/pkg/errors.TestNew.func1()
      /go/src/github.com/wego/pkg/errors/error_test.go:32 +0xcc

Previous write at 0x00c0000bcff8 by goroutine 18:
  github.com/wego/pkg/errors.(*Error).propagateContexts()
      /go/src/github.com/wego/pkg/errors/error.go:202 +0x6b8
  github.com/wego/pkg/errors.New()
      /go/src/github.com/wego/pkg/errors/error.go:77 +0x478
  github.com/wego/pkg/errors.TestNew.func1()
      /go/src/github.com/wego/pkg/errors/error_test.go:32 +0xcc

Goroutine 19 (running) created at:
  github.com/wego/pkg/errors.TestNew()
      /go/src/github.com/wego/pkg/errors/error_test.go:30 +0xbc
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1648 +0x40

Goroutine 18 (running) created at:
  github.com/wego/pkg/errors.TestNew()
      /go/src/github.com/wego/pkg/errors/error_test.go:30 +0xbc
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1595 +0x194
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1648 +0x40
==================
--- FAIL: TestNew (0.00s)
    testing.go:1465: race detected during execution of test
FAIL
coverage: 46.3% of statements
FAIL	github.com/wego/pkg/errors	0.441s
FAIL
make: *** [test] Error 1
```

After fixing the code:

```sh
$ make test
go test ./... -race -cover -p 5
ok  	github.com/wego/pkg/errors	1.505s	coverage: 46.3% of statements
```